### PR TITLE
enhance: Optimize channel node balancing for uneven QN distribution

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #42860
Fix channel node allocation when QueryNode count is not a multiple of channel count. The previous algorithm used simple division which caused uneven distribution with remainders.

Key improvements:
- Implement smart remainder distribution algorithm
- Refactor large function into focused helper functions
- Support two-phase rebalancing (release then allocate)
- Handle edge cases like insufficient nodes gracefully

### CODE REVIEW

**Comment by weiliu1031** [comment]
rerun cpp-unit-test

**Comment by weiliu1031** [comment]
rerun cpp-unit-test

File: internal/querycoordv2/meta/replica.go [hunk 1]
```diff
@@ -373,54 +381,209 @@ func (replica *mutableReplica) removeChannelExclusiveNodes(nodes ...int64) {
 	}
 }
 
+func (replica *mutableReplica) TryEnableChannelExclusiveMode(channelNames ...string) {
```
Comment (tedxu) [inline internal/querycoordv2/meta/replica.go:384]: Can we just return the availability of exclusive mode, and remove the `IsChannelExclusiveModeEnabled` function? That would be much simpler.

File: internal/querycoordv2/meta/replica.go [hunk 2]
```diff
@@ -373,54 +381,209 @@ func (replica *mutableReplica) removeChannelExclusiveNodes(nodes ...int64) {
...
+
+	// Calculate optimal node assignments
+	targetAssignments := replica.calculateOptimalAssignments(channelNodeInfos)
+
+	// Apply the rebalancing with minimal node movement
+	replica.rebalanceChannelNodes(channelNodeInfos, targetAssignments)
+}
+
+// getBalanceConfig retrieves and caches balance configuration
+func (replica *mutableReplica) getBalanceConfig() balanceConfig {
```
Comment (tedxu) [inline internal/querycoordv2/meta/replica.go]: nit: returning pointer instead of value.

Comment (tedxu) [reply internal/querycoordv2/meta/replica.go]: The struct `balanceConfig` looks not very useful, in the other hand, it adds additional complexity and memory overhead.

File: internal/querycoordv2/meta/replica.go [hunk 3]
```diff
@@ -373,54 +381,209 @@ func (replica *mutableReplica) removeChannelExclusiveNodes(nodes ...int64) {
 	}
 }
 
+func (replica *mutableReplica) TryEnableChannelExclusiveMode(channelNames ...string) {
+	if replica.replicaPB.ChannelNodeInfos == nil {
+		replica.replicaPB.ChannelNodeInfos = make(map[string]*querypb.ChannelNodeInfo)
+		for _, channelName := range channelNames {
+			replica.replicaPB.ChannelNodeInfos[channelName] = &querypb.ChannelNodeInfo{}
```
Comment (tedxu) [inline internal/querycoordv2/meta/replica.go:388]: Do we need to update this field if there is a node reconfiguration? E.g., when several nodes are brought down and the new configuration may not meet the exclusive requirements.

File: internal/querycoordv2/meta/replica.go [hunk 4]
```diff
@@ -373,54 +381,209 @@ func (replica *mutableReplica) removeChannelExclusiveNodes(nodes ...int64) {
...
+}
+
+// balanceConfig holds the configuration for channel balancing
+type balanceConfig struct {
+	EnableChannelExclusiveMode bool
+	channelExclusiveFactor     int
+}
+
+// channelNodeAssignment represents a channel's node assignment
+type channelNodeAssignment struct {
```
Comment (tedxu) [inline internal/querycoordv2/meta/replica.go]: nit: declare this in function `getSortedChannelsByNodeCount`
